### DISCUSSION TITLE
How to configure 2nd nginx (openresty) for ssl proxying?

### DISCUSSION THREAD

**Discussion by S-Przybylski** [blakeblackshear/frigate#845] [Ask A Question]
Hi
what should I do to enhance a running Nginx Proxy Manager (not on hass os) to access Frigate?
Adding a simple additional location "/frigate" --> http://IP-Frigate:5000 besides Home Assistant doesn't help at all.

Wanted scenario:
Client --> https://IP-Synology:4443/frigate --> http://IP-Frigate:5000

Environment:
I use Frigate on a Raspi (IP-Frigate) and Home Assistant on Docker on Synology (IP-Synology). 
In addition to that I have installed Nginx Proxy Manager on Synology (Docker) to provide interally SSL
I have created local certificates and imported them.
Home Assistant runs well.
I do not use subdomains.
No external connection defined.

**Comment by blakeblackshear** [comment]
Easiest way is to use a subdomain. If you can't do that, you can use nginx to inject a base url via the `x-ingress-path` header.

**Reply by S-Przybylski** [reply]
Hi @blakeblackshear 
I can't use a subdomain. I searched the docu of nginx (not a kubernetes one/its an openresty with web console on docker host network) for x-ingress-path. But I can't find anything about.
What I see is (browser debug), that the third request is not updated with the "leading" path "/frigate" in my configuration:
1st: https://synology-ip:4443/frigate 200
2nd:  https://synology-ip:4443/frigate/ 200
3rd: https://synology-ip:4443/dist/index.567mFe_hL2UA.js 404

The 3rd Get point to the HA web server then (default) and not to the Frigate server. The url wrongly begins with /dist instead of /frigate/dist. It seems that the replacement do not work.
Any idea about that? 

Configuration:
  location /frigate/ {
    proxy_set_header Host $host;
    proxy_set_header X-Forwarded-Scheme $scheme;
    proxy_set_header X-Forwarded-Proto  $scheme;
    proxy_set_header X-Forwarded-For    $remote_addr;
    proxy_pass       http://frigate-ip:5000/;
proxy_http_version 1.1;
proxy_set_header Upgrade $http_upgrade;
proxy_set_header Connection "upgrade";
  }

**Reply by blakeblackshear** [reply]
You need to set the header like this:

```
location /frigate/ {
  proxy_set_header Host $host;
  proxy_set_header X-Ingress-Path "/frigate";
  proxy_set_header X-Forwarded-Scheme $scheme;
  proxy_set_header X-Forwarded-Proto $scheme;
  proxy_set_header X-Forwarded-For $remote_addr;
  proxy_pass http://frigate-ip:5000/;
  proxy_http_version 1.1;
  proxy_set_header Upgrade $http_upgrade;
  proxy_set_header Connection "upgrade";
}
```

**Reply by S-Przybylski** [reply]
I only have one cosmetic issue now: If I navigate within Frigate, then the seen URL is the wrong: Its the one without the "/frigate". The navigation works as expected. A reload fails.
Is there also a Proxy setting available ?

**Reply by blakeblackshear** [reply]
I'm not sure what you mean. Can you be more specific?

**Reply by S-Przybylski** [reply]
Dear @blakeblackshear 
If I open the SSL link using https://synology-ip:4443/frigate the frigate webpage appear. You can navigate and use it. Everything works as expected. For example if you click on the first picture (cam1), then the next sub menu is loaded incl. the picture but the visible URL changes to https://synology-ip:4443/cameras/cam1. 
If you then reload the url by pressing F5 the browser tries to reach out e.g. https://synology-ip:4443/cameras/cam1 which is not reachable (it should be https://synology-ip:4443/frigate/cameras/cam1), because the reverse proxy sends this request to HA (default location).
Could it be that all links (buttons and pictures) of the web side are not relative ones?

**Reply by blakeblackshear** [reply]
The header you added is supposed to update all the urls in all files.

**Reply by S-Przybylski** [reply]
Here is an example:
1. I open a browser https//synology-ip/4443/frigate # this is the openresty nginx proxy
2. Then i point my cursor to the Frigate symbol. At the lower left corner you can see the wrong url then. If i click on events (not shown in the picture), the event page opens, but the browser url is wrong (without /frigate).

![grafik](https://user-images.githubusercontent.com/45458593/109399871-af8eae80-7945-11eb-8ad1-89673e7062e6.png)
![grafik](https://user-images.githubusercontent.com/45458593/109400250-ee256880-7947-11eb-9970-711e631f2d6f.png)

* Debug view: 
https://10.10.0.21:4443/frigate/api/events?include_thumbnails=0&limit=25&
Header:
Anforderungs-URL: https://10.10.0.21:4443/frigate/api/events?include_thumbnails=0&limit=25&
Anforderungsmethode: GET
Statuscode: 200 
Remote-Adresse: 10.10.0.21:4443
Referrer-Richtlinie: strict-origin-when-cross-origin
access-control-allow-origin: *
cache-control: no-store
content-encoding: gzip
content-type: application/json
date: Sat, 27 Feb 2021 20:53:01 GMT
server: openresty
vary: Accept-Encoding
:authority: 10.10.0.21:4443
:method: GET
:path: /frigate/api/events?include_thumbnails=0&limit=25&
:scheme: https
accept: */*
accept-encoding: gzip, deflate, br
accept-language: de,de-DE;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6
cookie: <deleted>
referer: https://10.10.0.21:4443/events
sec-fetch-dest: empty
sec-fetch-mode: cors
sec-fetch-site: same-origin
user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.182 Safari/537.36 Edg/88.0.705.81
include_thumbnails: 0
limit: 25
(leer)

Used Config:
  location /frigate/ {
    proxy_set_header Host $host;
    proxy_set_header X-Forwarded-Scheme $scheme;
    proxy_set_header X-Forwarded-Proto  $scheme;
    proxy_set_header X-Forwarded-For    $remote_addr;
    proxy_pass       http://10.10.0.41:5000/;
    proxy_set_header X-Ingress-Path "/frigate";
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
  }

**Reply by S-Przybylski** [reply]
Dear @blakeblackshear 
any idea about that?

**Reply by S-Przybylski** [reply]
Thats realy strange! I found some docs, which state that in some cases its needed to additional actions in addition for the way back (option4): http://sawers.com/blog/reverse-proxying-with-nginx/
I haven't tested that and I'am not sure if the Nginx Proxy Manager has that module built in.

**Reply by S-Przybylski** [reply]
What I found out is, that href are not updated as expected:
Example: 
````
<a class="block p-2 text-sm font-semibold text-gray-900 rounded hover:bg-blue-500 dark:text-gray-200 hover:text-white dark:hover:text-white focus:outline-none ring-opacity-50 focus:ring-2 ring-blue-300" href="/events"><div>Events</div></a>
````
I think it should be href="/frigate/events", href="./events", or ?

**Reply by blakeblackshear** [reply]
@paularmstrong any thoughts on why this would behave differently than it does for the addon?

**Reply by paularmstrong** [reply]
Ah yeah. The routing inside the application is not baseURL aware. This should be fixable.

**Reply by paularmstrong** [reply]
The addon will always refresh at the `/` path of the webapp, so that's why you won't see any issues.

**Reply by thehijacker** [reply]
> You need to set the header like this:
> 
> ```
> location /frigate/ {
>   proxy_set_header Host $host;
>   proxy_set_header X-Ingress-Path "/frigate";
>   proxy_set_header X-Forwarded-Scheme $scheme;
>   proxy_set_header X-Forwarded-Proto $scheme;
>   proxy_set_header X-Forwarded-For $remote_addr;
>   proxy_pass http://frigate-ip:5000/;
>   proxy_http_version 1.1;
>   proxy_set_header Upgrade $http_upgrade;
>   proxy_set_header Connection "upgrade";
> }
> ```

This is working for base main url. But when you wish to open events it redirects to https://domain/events and not on https://domain/frigate/events.

Any idea how to solve this?

Thank you.

**Reply by blakeblackshear** [reply]
The X-ingress-path header must not be making it to frigate somehow. That is what tells frigate to rewrite all the paths in the html, etc.

**Reply by simeon-walker** [reply]
I'm just trying this and there's something a little odd shows up. I have basically the same config as above.

When I go to the /frigate url I do get the page, however, when I click on a link such as Events, Cameras etc then the new page loads but the url shown in the browser is /events rather than /frigate/events

I'm not a JS programmer but I suspect the page is changing the url in response to the click but not taking the /frigate path into account. Since it's not a plain old page load to the browser, it looks wrong but did work.

Pressing reload on such a page will fail of course.

**Reply by thehijacker** [reply]
Same here but until I press reload all clicks are working. Done something better. Installed Wireguard VPN. Now can access all via local IP wnd port. Using Homer dashboard to bookmark all local IP addresses and port. Best combination ever.

**Reply by hack2spider** [reply]
I believ best thing is to run frigate on itself on a separate machine .. like this you can reach it with nginx ...

**Reply by simeon-walker** [reply]
It doesn't matter if you have a separate machine or where the nginx proxy runs. The issue is the way in which the web application handles the path through which it is accessed via the proxy.

**Reply by naanlizard** [reply]
This is still an issue, three years later. Any idea how to fix the URLs not updating properly to e.g. `hostname.tld/frigate`?

**Comment by mitchross** [comment]
I wrote an article on this https://mitchross09.medium.com/how-i-self-host-my-own-sites-and-applications-with-unraid-docker-authelia-and-cloudflare-1f0d8e6f8912

**Comment by ves1120** [comment]
This works for me. Set `X-Forwarded-For` to `$proxy_add_x_forwarded_for` instead of `$remote_addr` did it. Keeps the '/frigate' prefix even if you click on other links.
```
    server {
        location /frigate/ {
            proxy_pass http://frigate-ip:5000/;
            proxy_set_header Host $host;
            proxy_set_header X-Ingress-Path "/frigate";
            proxy_set_header X-Forwarded-Host $host:$server_port;
            proxy_set_header X-Forwarded-Server $host;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header X-Forwarded-Scheme $scheme;
            proxy_set_header X-Forwarded-Proto $scheme;
            proxy_http_version 1.1;
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection "upgrade";
        }
    }
```
Before adding `X-Ingress-Path`, I was getting forwarded to a blank page or Bad Gateway error. Thanks [blakeblackshear](https://github.com/blakeblackshear)
